File: examples/hello_lifecycle/hello_lifecycle_py/hello_lifecycle_py/hello_lifecycle_publisher.py
#!/usr/bin/env python3

'''
    A minimal ROS2 python node that publishes a string every second.

    Basic structure for ROS2 node publisher:
    - Initialise rclpy
    - Create a node
    - Create the publisher when in inactive state
    - Create a timer to publish at specific intervals when in active state
    - Keep it alive with spin
    - Shutdown

    From inactive to active, call the service lifecycle_activate with SetBool.Request true
    From active to inactive, call the service lifecycle_activate with SetBool.Request false

    BEST_EFFORT and VOLATILE have been selected for the QOS_PROFILE, as the
    node keeps publishing at fixed intervals and losing messages it's fine.

    * If you want reliability change BEST_EFFORT to RELIABLE.
    * If you want persisting samples for late-joining subscriptions,
      change VOLATILE to TRANSIENT_LOCAL.

    Author: Ioannis Selinis <selinis.g@gmail.com> 2025
'''

import sys
import asyncio

import rclpy
from rclpy.lifecycle import LifecycleNode, LifecycleState, TransitionCallbackReturn
from rclpy.executors import SingleThreadedExecutor
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from example_interfaces.msg import String
from example_interfaces.srv import SetBool


# https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html
QOS_PROFILE = QoSProfile(
    reliability=ReliabilityPolicy.BEST_EFFORT,      # RELIABLE or BEST_EFFORT
    durability=DurabilityPolicy.VOLATILE,           # VOLATILE or TRANSIENT_LOCAL
    history=HistoryPolicy.KEEP_LAST,                # KEEP_ALL or KEEP_LAST
    depth=1                                         # Used only with KEEP_LAST
)
QOS_PROFILE_SRV = QoSProfile(
    reliability=ReliabilityPolicy.RELIABLE,         # RELIABLE or BEST_EFFORT
    durability=DurabilityPolicy.VOLATILE,           # VOLATILE or TRANSIENT_LOCAL
    history=HistoryPolicy.KEEP_LAST,                # KEEP_ALL or KEEP_LAST
    depth=1                                         # Used only with KEEP_LAST
)
NODE_NAME = 'hello_lifecycle_publisher'


class HelloLifecyclePublisher(LifecycleNode):
    """The HelloLifecyclePublisher class that inherits from rclpy.node.Node.
    Initialises with a counter and a publisher and creates a timer
    that triggers once per second to publish a message.
    """

    def __init__(self):
        super().__init__(node_name=NODE_NAME)

        self.counter = 0
        self.cancel = False
        self.publisher = None
        self.timer = None
        self.server = None
        self.current_state = 'Unconfigured'

    def publisher_callback(self):
        """Publish the message once per second."""
        msg = String()
        msg.data = f'Lifecycle counter: {self.counter}'
        self.publisher.publish(msg=msg)
        self.counter += 1

    def service_callback(self, request: SetBool.Request, response: SetBool.Response) -> SetBool.Response:
        """Put the node in activate or deactivate mode based on the request flag."""
        self.get_logger().info(f'Service Callback is triggered with {request.data}')

        response.success = True
        self.get_logger().info(f'Current state is {self.current_state}')
        if self.current_state.lower() == 'unconfigured':
            # state is unconfigured:
            # i) go to inactive if request.data is True
            # ii) go to finalized if request.data is False
            if request.data:
                response.message = 'Node configured'
                self.trigger_configure()
            else:
                response.message = 'Node finalized (shutdown)'
                self.trigger_shutdown()
        elif self.current_state.lower() == 'inactive':
            # state is inactive:
            # i) go to active if request.data is True
            # ii) go to unconfigured if request.data is False
            # iii) go to finalized (we don't consider it here due to SetBool states)
            if request.data:
                self.trigger_activate()
                response.message = 'Node activated'
            else:
                self.trigger_cleanup()
                response.message = 'Node unconfigured (cleaned up)'
        elif self.current_state.lower() == 'active':
            # state is active:
            # i) go to shutdown if request.data is True
            # ii) go to inactive if request.data is False
            if request.data:
                response.message = 'Node finalized (shutdown)'
                self.trigger_shutdown()
            else:
                response.message = 'Node deactivated'
                self.trigger_deactivate()
        return response

    def on_configure(self, state: LifecycleState) -> TransitionCallbackReturn:
        """Set the node state to configure (create the publisher)."""
        self.current_state = 'Inactive'
        self.get_logger().info(f'{NODE_NAME} transitions from {state.label} ({state.state_id}) to {self.current_state} - create the publisher & service')
        self.publisher = self.create_publisher(msg_type=String,
                                               topic='hello_lifecycle',
                                               qos_profile=QOS_PROFILE)
        self.server = self.create_service(srv_type=SetBool,
                                          srv_name='lifecycle_activate',
                                          callback=self.service_callback,
                                          qos_profile=QOS_PROFILE_SRV)
        return TransitionCallbackReturn.SUCCESS

    def on_activate(self, state: LifecycleState) -> TransitionCallbackReturn:
        """Set the node state to activate (create the timer for the publisher)."""
        self.current_state = 'Active'
        self.get_logger().info(f'{NODE_NAME} transitions from {state.label} ({state.state_id}) to {self.current_state} - create the timer')
        self.timer = self.create_timer(timer_period_sec=1.0,
                                       callback=self.publisher_callback)
        return super().on_activate(state)

    def on_cleanup(self, state: LifecycleState) -> TransitionCallbackReturn:
        """Cleanup state"""
        self.current_state = 'Unconfigured'
        self.get_logger().info(f'{NODE_NAME} transitions from {state.label} ({state.state_id}) to {self.current_state} -- cleanup')
        self.destroy_lifecycle_publisher(self.publisher)
        self.destroy_timer(self.timer)
        return TransitionCallbackReturn.SUCCESS

    def on_deactivate(self, state: LifecycleState) -> TransitionCallbackReturn:
        """Deactivate node."""
        self.current_state = 'Inactive'
        self.get_logger().info(f'{NODE_NAME} transitions from {state.label} ({state.state_id}) to {self.current_state} -- deactivate timer')
        if self.timer:
            self.timer.cancel()
        return super().on_deactivate(state)

    def on_shutdown(self, state: LifecycleState) -> TransitionCallbackReturn:
        """Shutdown node."""
        self.current_state = 'Finalized'
        self.get_logger().info(f'{NODE_NAME} transitions from {state.label} ({state.state_id}) to {self.current_state} -- shutdown')
        self.destroy_lifecycle_publisher(self.publisher)
        self.destroy_timer(self.timer)
        self.cancel = True
        return super().on_shutdown(state)


async def spin(executor: SingleThreadedExecutor, node: HelloLifecyclePublisher):
    """Async spin."""
    while not node.cancel:
        executor.spin_once(timeout_sec=0.1)
        # Setting the delay to 0 provides an optimized path to allow other tasks to run.
        await asyncio.sleep(0)


def main(args=None) -> None:
    """Entry Point for our node"""
    try:
        rclpy.init(args=args)
        executor = SingleThreadedExecutor()
        node = HelloLifecyclePublisher()
        executor.add_node(node=node)
        # Node state: UNCONFIGURED --> INACTIVE
        node.trigger_configure()

        loop = asyncio.get_event_loop()
        loop.run_until_complete(spin(executor=executor, node=node))
    except Exception as err:  # pylint: disable=broad-except
        node.get_logger().error(f'Main loop crashed: {err}')
    finally:
        node.get_logger().error(f'Shutting down {NODE_NAME}')
        node.destroy_node()
        rclpy.shutdown()
        loop.close()
        sys.exit(0)


if __name__ == '__main__':
    main()

File: examples/hello_lifecycle/hello_lifecycle_cpp/src/hello_lifecycle_publisher.cpp
/**
 *  A minimal ROS2 cpp node that publishes a message every second.
 *
 *  Basic structure for ROS2 node:
 *  - Initialise rclcpp
 *  - Create a node
 *  - Create the publisher
 *  - Create a timer to publish at specific intervals
 *  - Keep it alive with spin
 *  - Shutdown
 *
 *  BEST_EFFORT and VOLATILE have been selected for the QOS_PROFILE, as the
 *  node keeps publishing at fixed intervals and losing messages it's fine.
 *
 *  * If you want reliability change BEST_EFFORT to RELIABLE.
 *  * If you want persisting samples for late-joining subscriptions,
 *    change VOLATILE to TRANSIENT_LOCAL.
 *
 *  Author: Ioannis Selinis <selinis.g@gmail.com> 2025
 */

#include <rclcpp/rclcpp.hpp>
#include <rclcpp/executors/single_threaded_executor.hpp>
#include <rclcpp_lifecycle/lifecycle_node.hpp>
#include "example_interfaces/msg/string.hpp"
#include "example_interfaces/srv/set_bool.hpp"


static const rmw_qos_profile_t rmw_qos_profile =
{
    RMW_QOS_POLICY_HISTORY_KEEP_LAST,
    1,
    RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT,
    RMW_QOS_POLICY_DURABILITY_VOLATILE,
    RMW_QOS_DEADLINE_DEFAULT,
    RMW_QOS_LIFESPAN_DEFAULT,
    RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
    RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
    false
};


class HelloLifecyclePublisher : public rclcpp_lifecycle::LifecycleNode
{
    public:
        HelloLifecyclePublisher() : LifecycleNode("hello_lifecycle_c_publisher")
        {
            m_counter = 0;
            m_executor = nullptr;
            RCLCPP_INFO(this->get_logger(), "Node has started");
        }

        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
        on_configure(const rclcpp_lifecycle::State &state) override
        {
            auto qos = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile), rmw_qos_profile);
            m_current_state = "inactive";
            RCLCPP_INFO(get_logger(), "%s transitions from %s (%d) to %s - create publisher & service", get_name(), state.label().c_str(), state.id(), m_current_state.c_str());
            m_publisher = this->create_publisher<example_interfaces::msg::String>("hello_lifecycle", qos);

            m_server = this->create_service<example_interfaces::srv::SetBool>("lifecycle_activate",
                                                                              std::bind(&HelloLifecyclePublisher::serviceCallback,
                                                                                        this, std::placeholders::_1,
                                                                                        std::placeholders::_2));
            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
        }
        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
        on_activate(const rclcpp_lifecycle::State &state) override
        {
            m_current_state = "active";
            RCLCPP_INFO(get_logger(), "%s transitions from %s (%d) to %s", get_name(), state.label().c_str(), state.id(), m_current_state.c_str());
            m_publisher->on_activate();
            m_timer = this->create_wall_timer(std::chrono::seconds(1),
                                              std::bind(&HelloLifecyclePublisher::timerCallback, this));
            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
        }
        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
        on_deactivate(const rclcpp_lifecycle::State &state) override
        {
            m_current_state = "inactive";
            RCLCPP_INFO(get_logger(), "%s transitions from %s (%d) to %s", get_name(), state.label().c_str(), state.id(), m_current_state.c_str());
            m_publisher->on_deactivate();
            m_timer->reset();
            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
        }
        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
        on_cleanup(const rclcpp_lifecycle::State &state) override
        {
            m_current_state = "unconfigured";
            RCLCPP_INFO(get_logger(), "%s transitions from %s (%d) to %s", get_name(), state.label().c_str(), state.id(), m_current_state.c_str());
            m_publisher->on_deactivate();
            m_timer->reset();
            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
        }
        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
        on_shutdown(const rclcpp_lifecycle::State &state) override
        {
            m_current_state = "finalized";
            RCLCPP_INFO(get_logger(), "%s transitions from %s (%d) to %s", get_name(), state.label().c_str(), state.id(), m_current_state.c_str());
            m_publisher->on_deactivate();
            m_timer->reset();
            if (m_executor) {
                m_executor->cancel();
            }
            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
        }
        void set_executor(rclcpp::executors::SingleThreadedExecutor* executor)
        {
            m_executor = executor;
        }
    private:
        void timerCallback()
        {
            auto msg = example_interfaces::msg::String();
            msg.data = "Lifecycle counter: " + std::to_string(m_counter);
            m_publisher->publish(msg);
            m_counter++;
        }
        void serviceCallback(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                             const example_interfaces::srv::SetBool::Response::SharedPtr response)
        {
            RCLCPP_INFO(get_logger(), "Service callback triggered with %s", request->data ? "true" : "false");
            if (m_current_state == ""){
                response->success = false;
                response->message = "Unknown state";
                return;
            }
            response->success = true;
            if (m_current_state == "unconfigured"){
                if (request->data){
                    trigger_transition(lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE);
                    response->message = "Node configured";
                } else{
                    trigger_transition(lifecycle_msgs::msg::Transition::TRANSITION_UNCONFIGURED_SHUTDOWN);
                    response->message = "Node finalized";
                }
            } else if (m_current_state == "inactive") {
                if (request->data) {
                    trigger_transition(lifecycle_msgs::msg::Transition::TRANSITION_ACTIVATE);
                    response->message = "Node activated";
                } else {
                    trigger_transition(lifecycle_msgs::msg::Transition::TRANSITION_CLEANUP);
                    response->message = "Node cleaned up";
                }
            } else if (m_current_state == "active") {
                if (request->data) {
                    trigger_transition(lifecycle_msgs::msg::Transition::TRANSITION_ACTIVE_SHUTDOWN);
                    response->message = "Node finalized (shutdown)";
                } else {
                    trigger_transition(lifecycle_msgs::msg::Transition::TRANSITION_DEACTIVATE);
                    response->message = "Node deactivated";
                }
            }
        }
        rclcpp::executors::SingleThreadedExecutor* m_executor;
        rclcpp::TimerBase::SharedPtr m_timer;
        rclcpp_lifecycle::LifecyclePublisher<example_interfaces::msg::String>::SharedPtr m_publisher;
        rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr m_server;
        std::string m_current_state;
        int m_counter = 0;
};


int main (int argc, char **argv)
{
    try{
        rclcpp::init(argc, argv);
        rclcpp::executors::SingleThreadedExecutor executor;
        auto node = std::make_shared<HelloLifecyclePublisher>();
        node->set_executor(&executor);
        executor.add_node(node->get_node_base_interface());
        node->trigger_transition(lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE);
        executor.spin();
        rclcpp::shutdown();                             // Shutdown executor
    } catch (...){
        RCLCPP_ERROR(rclcpp::get_logger("hello_lifecycle_publisher"), "Main loop crashed");
    }
}
